community/index/src/main/java/org/neo4j/index/ValueMerger.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.index;
+
+/**
+ * Decides what to do when inserting key which already exists in index. Different implementations of
+ * {@link ValueMerger} can result in unique/non-unique indexes for example.
+ *
+ * @param <VALUE> type of values to merge.
+ */
+public interface ValueMerger<VALUE>
+{
+    /**
+     * Merge an existing value with a new value, returning potentially a combination of the two, or {@code null}
+     * if no merge was done effectively meaning that nothing should be written.
+     *
+     * @param value existing value
+     * @param withValue new value
+     * @return {@code value}, now merged with {@code withValue}, or {@code null} if no merge was done.
```

Comment:
We took inspiration from `Map#merge` where a `null` from provided `BiFunction` means remove entry from map. We can not do that because we have a different path for remove. So we want to explicitly define the expected behaviour for when `ValueMerger` returns null.

We made it like this to avoid confusion, so it actually creates more confusion, it should be changed. Does it?

Note: Returning the old value will result in a write with no effective change of the actual value. We could add a check to avoid writing in that case.